plugins/out_kafka/kafka.c

```diff
@@ -151,6 +156,35 @@ int produce_message(struct flb_time *tm, msgpack_object *map,
                 topic = flb_kafka_topic_lookup((char *) val.via.str.ptr,
                                                val.via.str.size,
                                                ctx);
+                /* Add topic on the fly to topiclist and use it */
+                if (ctx->dynamic_topic) {
+                    if (strncmp(topic->name, flb_kafka_topic_default(ctx)->name, val.via.str.size) == 0 &&
+                       (strncmp(topic->name, val.via.str.ptr, val.via.str.size) != 0) ) {
+                        dynamic_topic = flb_malloc(val.via.str.size + 1);
+                        if (!dynamic_topic) {
+                            flb_errno();
+                            /* in this case we use the default topic */
+                            break;
+                        }
+                        strncpy(dynamic_topic, val.via.str.ptr, val.via.str.size);
+                        dynamic_topic[val.via.str.size] = '\0';
+
+                        topics = flb_utils_split(dynamic_topic, ',', 1);
+                        mk_list_foreach(head, topics) {
+                            entry = mk_list_entry(head, struct flb_split_entry, _head);
+                            topic = flb_kafka_topic_create(entry->value, ctx);
+                            if (!topic) {
+                                flb_error("[out_kafka] cannot register topic '%s'",
+                                          entry->value);
+                            }
+                            else {
+                                flb_info("[out_kafka] new topic added: %s", dynamic_topic);
+                                break;
+                            }
+                        flb_free(dynamic_topic);
```

Comment:
not tested but, if ```!topic``` in line 176 fails, looks like you are freeing dynamic_topic as many entries are found in the topics linked list.

Reply:
made changes